### PR TITLE
remove checkBucketExist check entirely to avoid fan-out calls

### PR DESCRIPTION
**Pull request by harshavardhana**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
remove checkBucketExist check entirely to avoid fan-out calls

## Motivation and Context
Each Put, List, and Multipart operation heavily relies on 
on making a GetBucketInfo() call to verify whether a bucket 
exists regularly. This has a significant performance cost 
when there are many servers involved.

We did optimize this part by vectorizing the bucket calls; 
however, it's not enough beyond 100 nodes, which becomes 
reasonably visible in terms of performance.

## How to test this PR?
CI/CD has enough tests, its also changing few things that
need attention and careful review.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/erasure-server-pool.go [hunk 1]
```diff
@@ -1307,7 +1307,10 @@ func (z *erasureServerPools) ListObjectVersions(ctx context.Context, bucket, pre
 
 	merged, err := z.listPath(ctx, &opts)
 	if err != nil && err != io.EOF {
-		return loi, err
+		if !isErrBucketNotFound(err) {
+			logger.LogOnceIf(ctx, err, "erasure-list-objects-path-"+bucket)
```
Comment (klauspost) [inline cmd/erasure-server-pool.go:1311]: What is this for? Seems like a leftover.

File: cmd/metacache-server-pool.go [hunk 2]
```diff
@@ -313,6 +314,11 @@ func (z *erasureServerPools) listMerged(ctx context.Context, o listPathOptions,
 	}
 
 	if isAllNotFound(errs) {
+		for _, err := range errs {
+			if err != nil && errors.Is(err, errVolumeNotFound) {
+				return err
+			}
+		}
```
Comment (klauspost) [inline cmd/metacache-server-pool.go]: Potentially a risky change since it is made without quorum check.

File: cmd/metacache-set.go [hunk 3]
```diff
@@ -1083,6 +1084,11 @@ func listPathRaw(ctx context.Context, opts listPathRawOptions) (err error) {
 					errDiskNotFound.Error():
 					atEOF++
 					fnf++
+					// This is a special case, to handle bucket does
+					// not exist situations.
+					if errors.Is(err, errVolumeNotFound) {
+						volumeNotFound = err
+					}
```
Comment (klauspost) [inline cmd/metacache-set.go]: Maybe count up separately like fnf.

File: cmd/object-api-input-checks.go [hunk 4]
```diff
@@ -40,7 +40,7 @@ func checkDelObjArgs(ctx context.Context, bucket, object string) error {
 // Checks bucket and object name validity, returns nil if both are valid.
 func checkBucketAndObjectNames(ctx context.Context, bucket, object string) error {
 	// Verify if bucket is valid.
-	if !isMinioMetaBucketName(bucket) && s3utils.CheckValidBucketName(bucket) != nil {
+	if !isMinioMetaBucketName(bucket) && s3utils.CheckValidBucketNameStrict(bucket) != nil {
```
Comment (klauspost) [inline cmd/object-api-input-checks.go:43]: Why this change? Seems this could make existing buckets unreachable?

(same for similar change below)

**Review by klauspost — COMMENTED** [review]
Just some minor stuff.

File: cmd/object-api-input-checks.go (hunk 4 above)
Comment (harshavardhana) [reply cmd/object-api-input-checks.go:43]: Well, this was there for FS mode back in the days when we supported existing buckets; we never allowed the creation of new invalid buckets.

Look at the generic-handlers.go

```
                        // Validate bucket names if it is not empty
                        if bucketName != "" && s3utils.CheckValidBucketNameStrict(bucketName) != nil {
                                if ok {
                                        tc.FuncName = "handler.ValidRequest"
                                        tc.ResponseRecorder.LogErrBody = true
                                }
                                defer logger.AuditLog(r.Context(), w, r, mustGetClaimsFromToken(r))
                                writeErrorResponse(r.Context(), w, errorCodes.ToAPIErr(ErrInvalidBucketName), r.URL)
                                return
                        }
```

File: cmd/metacache-set.go (hunk 3 above)
Comment (harshavardhana) [reply cmd/metacache-set.go]: okay I can do  that

File: cmd/erasure-server-pool.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/erasure-server-pool.go:1311]: This I copied over from the ListObjects() code @klauspost

File: cmd/metacache-server-pool.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/metacache-server-pool.go]: Okay

**Comment by harshavardhana** [comment]
@klauspost fixed some of your comments

**Review by klauspost — APPROVED** [review]

File: cmd/erasure-multipart.go [hunk 5]
```diff
@@ -582,6 +580,13 @@ func writeAllDisks(ctx context.Context, disks []StorageAPI, dstBucket, dstEntry
 func (er erasureObjects) PutObjectPart(ctx context.Context, bucket, object, uploadID string, partID int, r *PutObjReader, opts ObjectOptions) (pi PartInfo, err error) {
 	auditObjectErasureSet(ctx, object, &er)
 
+	data := r.Reader
+	// Validate input data size and it can never be less than zero.
+	if data.Size() < -1 {
```
Comment (fwessels) [inline cmd/erasure-multipart.go:585]: Shouldn't it be `<= -1` or simply `< 0` ?

Comment (harshavardhana) [reply cmd/erasure-multipart.go:585]: no -1 is valid for compressed stream @fwessels

**Review by fwessels — APPROVED** [review]
LGTM

**Review by vadmeste — APPROVED** [review]
LGTM
